File: android/app/src/main/java/com/foodsense/android/services/APIService.kt
package com.foodsense.android.services

import android.graphics.Bitmap
import com.foodsense.android.data.FoodLog
import com.foodsense.android.data.MealPlanDay
import com.foodsense.android.data.NutritionInfo
import com.foodsense.android.data.QuizQuestion
import com.foodsense.android.data.UserStats
import com.foodsense.android.data.WeeklyInsight
import com.foodsense.android.services.ai.AIProviderManager

// Thin facade that preserves the `app.apiService.xxx()` call convention used
// by all consumer screens. Internally delegates every call to the active
// AIProvider selected by AIProviderManager.
//
// Consumer screens do NOT need to change — they keep calling
// `apiService.analyzeFood(bitmap)` etc.
class APIService {

    private val provider
        get() = AIProviderManager.instance?.activeProvider
            ?: throw IllegalStateException(
                "No AI provider configured. Add a Gemini API key or download the on-device model in Settings."
            )

    suspend fun analyzeFood(image: Bitmap): Pair<String, NutritionInfo> =
        provider.analyzeFood(image)

    suspend fun searchFood(query: String): Pair<String, NutritionInfo> =
        provider.searchFood(query)

    suspend fun getCoachAdvice(
        userStats: UserStats?,
        logs: List<FoodLog>,
        healthData: String,
        historyToon: String,
        userQuery: String,
    ): String = provider.getCoachAdvice(userStats, logs, healthData, historyToon, userQuery)

    suspend fun generateMealPlan(
        userStats: UserStats?,
        calorieBudget: Int,
    ): List<MealPlanDay> = provider.generateMealPlan(userStats, calorieBudget)

    suspend fun getWeeklyInsights(
        userStats: UserStats?,
        history: String,
        healthData: String,
    ): WeeklyInsight = provider.getWeeklyInsights(userStats, history, healthData)

    suspend fun getQuizQuestion(): QuizQuestion =
        provider.getQuizQuestion()

    suspend fun predictWeight(
        userStats: UserStats?,
        history: String,
    ): String = provider.predictWeight(userStats, history)
}

File: FoodDetectionApp/APIService.swift
import Foundation
import UIKit

/// Thin facade that preserves the `APIService.shared.xxx()` call convention
/// used by all consumer views. Internally delegates every call to the active
/// `AIProvider` selected by `AIProviderManager`.
///
/// Consumer views do NOT need to change — they keep calling
/// `APIService.shared.analyzeFood(image:)` etc.
class APIService {
    static let shared = APIService()

    private var provider: AIProvider {
        get throws {
            guard let p = AIProviderManager.shared.activeProvider else {
                throw NSError(
                    domain: "AIProviderError", code: -1,
                    userInfo: [NSLocalizedDescriptionKey: "No AI provider configured. Add a Gemini API key or download the on-device model in Settings."]
                )
            }
            return p
        }
    }

    private init() {}

    // MARK: - Food Analysis

    func analyzeFood(image: UIImage) async throws -> (name: String, info: NutritionInfo) {
        try await provider.analyzeFood(image: image)
    }

    // MARK: - Food Search (Text)

    func searchFood(query: String) async throws -> (name: String, info: NutritionInfo) {
        try await provider.searchFood(query: query)
    }

    // MARK: - AI Coach

    func getCoachAdvice(userStats: UserStats?, logs: [NutritionManager.FoodLog], healthData: String, historyTOON: String, userQuery: String) async throws -> String {
        try await provider.getCoachAdvice(userStats: userStats, logs: logs, healthData: healthData, historyTOON: historyTOON, userQuery: userQuery)
    }

    // MARK: - Meal Plan

    func generateMealPlan(userStats: UserStats?, calorieBudget: Int) async throws -> [PlannedMeal] {
        try await provider.generateMealPlan(userStats: userStats, calorieBudget: calorieBudget)
    }

    // MARK: - Portion Estimation

    func estimatePortion(image: UIImage) async throws -> (name: String, info: NutritionInfo, estimatedGrams: Double) {
        try await provider.estimatePortion(image: image)
    }

    // MARK: - Before/After Comparison

    func compareBeforeAfter(before: UIImage, after: UIImage) async throws -> String {
        try await provider.compareBeforeAfter(before: before, after: after)
    }

    // MARK: - OCR Nutrition Label

    func ocrNutritionLabel(image: UIImage) async throws -> (name: String, info: NutritionInfo) {
        try await provider.ocrNutritionLabel(image: image)
    }

    // MARK: - Generate Insights

    func generateInsights(foodHistory: String, calorieBudget: Int, userStats: UserStats?) async throws -> String {
        try await provider.generateInsights(foodHistory: foodHistory, calorieBudget: calorieBudget, userStats: userStats)
    }
}
